### PR TITLE
convert Rect to immutable and update popup positioning

### PR DESCRIPTION
**Pull request by dshovchko**

### CODE REVIEW

File: site/src/esl-popup/esl-d-popup-game.ts [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+  @memoize()
+  get $trigger(): HTMLElement | undefined {
+    if (this.trigger) return ESLTraversingQuery.first(this.trigger, this) as HTMLElement | undefined;
+  }
+
+  protected update(): void {
+    if (!this.$trigger) return;
+    if (this.dragging) this._onDragStop();
+    this.$$off(this._onDragStart);
+    this.$$on(this._onDragStart);
```
Comment (ala-n) [inline site/src/esl-popup/esl-d-popup-game.ts:25]: What the purpose for resubscription of the unconditionl, self-applied (target can not be changed) listener ?

Comment (dshovchko) [reply site/src/esl-popup/esl-d-popup-game.ts]: This is redundant code. It was the first attempt to construct a component. Probably I missed it to remove.
Thanks a lot!

File: src/modules/esl-popup/core/esl-popup-position.ts [hunk 2]
```diff
@@ -148,72 +158,75 @@ function adjustAlongMajorAxis(cfg: PopupPositionConfig, value: PopupPositionValu
...
-  let arrowLimit = cfg.marginArrow;
-  let coordAdjust = adjustAlignmentBySide(popup[start] - cfg.outer[start], arrow[start], arrowLimit, true);
-  if (coordAdjust) {
-    popup[start] -= coordAdjust;
-    arrow[start] += coordAdjust;
+  let diff = isStarting ? popup[start] - cfg.outer[start] : popup[end] - cfg.outer[end];
+  if (isWider) {
+    // apply another value on case when popup is greater than the outer container
+    diff = cfg.isRTL ? popup[end] - cfg.outer[end] : popup[start] - cfg.outer[start];
   }
```
Comment (ala-n) [inline src/modules/esl-popup/core/esl-popup-position.ts:222]: One more cosmetic (but immutable) idea
```suggestion
  const diffSide = (isWider ? !cfg.isRTL : isStarting) ? start : end;
  const diff = popup[diffSide] - cfg.outer[diffSide];
```

**Review by ala-n — APPROVED** [review]

File: src/modules/esl-popup/core/esl-popup-position.ts (hunk 2 above)
Comment (dshovchko) [reply src/modules/esl-popup/core/esl-popup-position.ts]: suggestion was applied with explanation comment

File: src/modules/esl-popup/core/esl-popup-position.ts [hunk 3]
```diff
@@ -148,72 +158,75 @@ function adjustAlongMajorAxis(cfg: PopupPositionConfig, value: PopupPositionValu
...
 
+/**
+ * Sets up the configuration for adjusting position along the minor axis
+ * @param cfg - popup position config
+ * @param popup - current popup's position value
+ * @returns configuration for adjusting position along the minor axis
+ */
+function setupAlignmentBySide(cfg: PopupPositionConfig, popup: Rect): PopupAlignmentConfig {
+  const isHorizontal = isOnHorizontalAxis(cfg.position);
+  const start = isHorizontal ? 'y' : 'x';
```
[the remark this answers is not in the document]
Comment (dshovchko) [reply src/modules/esl-popup/core/esl-popup-position.ts:191]: The problem with this is that
```ts
export interface Point {
  x: number;
  y: number;
}
```

File: src/modules/esl-popup/core/esl-popup-position.ts [hunk 4]
```diff
@@ -148,72 +158,76 @@ function adjustAlongMajorAxis(cfg: PopupPositionConfig, value: PopupPositionValu
...
-  arrowLimit += calcUsableSizeForArrow(cfg, dimension);
-  coordAdjust = adjustAlignmentBySide(popup[end] - cfg.outer[end], arrow[start], arrowLimit, false);
-  if (coordAdjust) {
-    popup[start] -= coordAdjust;
-    arrow[start] += coordAdjust;
-  }
-  return {...value, popup, arrow};
+  // the side for calculating the distance between the popup and the outer (container) bounding should be:
+  // - when the popup is wider than the container the diff side should depend on the text direction
+  //   (start side for LTR, end side for RTL)
```
Comment (yadamskaya) [inline src/modules/esl-popup/core/esl-popup-position.ts:220]:
```suggestion
  //  (start side for LTR, end side for RTL)
```

**Review by yadamskaya — APPROVED** [review]

**Review by Natalie-Smirnova — APPROVED** [review]

**Review by fshovchko — APPROVED** [review]

**Review by abarmina — APPROVED** [review]

File: src/modules/esl-popup/core/esl-popup-position.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (ala-n) [reply src/modules/esl-popup/core/esl-popup-position.ts:191]: But do we use it as a Point abstraction somewhere in here?
PS: just a question, PR still approved for merging and updates will be considered for later (if they actual)

[the remark this answers is not in the document]
Comment (ala-n) [reply src/modules/esl-popup/core/esl-popup-position.ts:191]: Nevermind, see that we use it to access `arrow` which  is the Point instance

**Comment by ala-n** [comment]
Maintenance note: corresponding commits description added for automated semantic releases

**Pull request merged**